### PR TITLE
feat(core): cluster role automatic management

### PR DESCRIPTION
**Pull request by zyxxoo**

### CODE REVIEW

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/MetaData.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    int epoch;
+
+    public MetaData(String node, int epoch) {
+        this.node = node;
+        this.epoch = epoch;
+        this.count = 1;
+    }
+
+    public void increaseCount() {
+        this.count ++;
```
Comment (imbajin) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/MetaData.java:18]: extra space

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/MetaData.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
+package com.baidu.hugegraph.election;
+
+import java.util.Objects;
+
+public class MetaData {
+
+    String node;
+    long count;
```
Comment (imbajin) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/MetaData.java:8]: do we need long?

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/MetaDataAdapter.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package com.baidu.hugegraph.election;
+
+import java.util.Optional;
+
+public interface MetaDataAdapter {
```
Comment (imbajin) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/MetaDataAdapter.java:5]: one more empty line

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/MetaData.java (hunk 2 above)
Comment (zyxxoo) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/election/MetaData.java:8]: maybe int is enough

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+
+import java.util.Optional;
+import java.util.concurrent.locks.LockSupport;
+
+import com.baidu.hugegraph.util.E;
+
+public class RoleElectionStateMachineImpl implements RoleElectionStateMachine {
+
+    private volatile boolean shutdown = false;
+    private Config config;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java:11]: mark final

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+
+        @Override
+        public Callback callback(StateMachineCallback callback) {
+            return callback::unknown;
+        }
+    }
+
+    private static class SafeState implements RoleState {
+
+        Integer epoch;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java:110]: mark private

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+
+        @Override
+        public Callback callback(StateMachineCallback callback) {
+            return callback::safe;
+        }
+    }
+
+    private static class MasterState implements RoleState {
+
+        MetaData metaData;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java:130]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+        }
+    }
+
+    @FunctionalInterface
+    private interface Callback {
+
+        void call(StateMachineContext context);
+    }
+
+    private static class UnKnownState implements RoleState {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java:74]: UnKnown => Unknown

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+
+        Integer epoch;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java:76]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/MetaData.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+        return this.epoch;
+    }
+
+    public long count() {
+        return this.count;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/MetaData.java:35]: expect to wrap line after if statement

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/MetaData.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+        if (!(o instanceof MetaData)) return false;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/MetaData.java:36]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/MetaData.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
+package com.baidu.hugegraph.election;
+
+import java.util.Objects;
+
+public class MetaData {
+
+    String node;
+    long count;
+    int epoch;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/MetaData.java:9]: mark private

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/MetaDataAdapter.java [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+
+    boolean postDelyIfPresent(MetaData metaData, long delySecond);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/MetaDataAdapter.java:7]: Dely => Delay?

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/MetaDataAdapter.java [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+
+    Optional<MetaData> queryDelay(long delySecond);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/MetaDataAdapter.java:9]: delySecond => delaySecond

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/RoleElectionStateMachineTest.java [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+            @Override
+            public void error(StateMachineContext context, Throwable e) {
+
+            }
+        };
+        final MetaDataAdapter adapter = new MetaDataAdapter() {
+            int epoch = 0;
+            int count = 0;
+            Map<Integer, MetaData> data = new ConcurrentHashMap<>();
+            @Override
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/RoleElectionStateMachineTest.java:173]: add a blank line

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/RoleElectionStateMachineTest.java [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+                if (logRecords.size() > MAX_COUNT) {
+                    context.stateMachine().shutdown();
+                }
+            }
+
+            @Override
+            public void error(StateMachineContext context, Throwable e) {
+
+            }
+        };
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/RoleElectionStateMachineTest.java:168]: add a blank line

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/Config.java [hunk 16]
```diff
@@ -0,0 +1,14 @@
+package com.baidu.hugegraph.election;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/Config.java:1]: add license header

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/MetaData.java [hunk 17]
```diff
@@ -0,0 +1,45 @@
+package com.baidu.hugegraph.election;
+
+import java.util.Objects;
+
+public class MetaData {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/MetaData.java:5]: can we rename to RoleStateData?

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachine.java [hunk 18]
```diff
@@ -0,0 +1,9 @@
+package com.baidu.hugegraph.election;
+
+public interface RoleElectionStateMachine {
+
+    void shutdown();
+
+    void apply(StateMachineCallback stateMachineCallback);
+
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachine.java:8]: delete the empty line

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java [hunk 19]
```diff
@@ -0,0 +1,281 @@
...
+                this.metaData = state.metaData;
+            } else {
+                this.count++;
+            }
+        }
+    }
+
+    private static class CandidateState implements RoleState {
+
+        Integer epoch;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java:205]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java [hunk 20]
```diff
@@ -0,0 +1,281 @@
...
+        }
+
+        public void merge(WorkerState state) {
+            if (state.metaData.epoch() > this.metaData.epoch()) {
+                this.count = 0;
+                this.metaData = state.metaData;
+            } else if (state.metaData.epoch() < this.metaData.epoch()){
+                throw new IllegalStateException("Epoch must increase");
+            } else if (state.metaData.epoch() == this.metaData.epoch() &&
+                    state.metaData.count() < this.metaData.count()) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java:191]: align with `state.metaData.epoch()`

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/MetaDataAdapter.java (hunk 13 above)
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/MetaDataAdapter.java:9]: prefer queryWithDelay name

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/MetaDataAdapter.java (hunk 12 above)
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/MetaDataAdapter.java:7]: prefer delayIfNodePresent

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java (hunk 7 above)
Comment (zyxxoo) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java:98]: UnKnown equal Unknown

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java [hunk 21]
```diff
@@ -9,15 +28,13 @@
 public class RoleElectionStateMachineImpl implements RoleElectionStateMachine {
 
     private volatile boolean shutdown = false;
-    private Config config;
-
+    private final Config config;
     private volatile RoleState state;
+    private final RoleStataDataAdapter roleStataDataAdapter;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java:33]: "Stata" -- typo

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleStataDataAdapter.java [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+ * under the License.
+ */
+
+package com.baidu.hugegraph.election;
+
+import java.util.Optional;
+
+public interface RoleStataDataAdapter {
+
+    boolean delayIfNodePresent(RoleStateData metaData, long delaySecond);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleStataDataAdapter.java:26]: updateIfNodePresent

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleStateData.java [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+ */
+
+package com.baidu.hugegraph.election;
+
+import java.util.Objects;
+
+public class RoleStateData {
+
+    private String node;
+    private long count;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleStateData.java:27]: rename to clock or times?

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/StateMachineCallback.java [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+
+    void master(StateMachineContext context);
+
+    void worker(StateMachineContext context);
+
+    void candidate(StateMachineContext context);
+
+    void unknown(StateMachineContext context);
+
+    void safe(StateMachineContext context);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/StateMachineCallback.java:32]: rename safe to abdication?

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/RoleElectionStateMachineTest.java [hunk 25]
```diff
@@ -0,0 +1,324 @@
...
+        for (LogEntry entry: logRecords) {
+            if (entry.role == LogEntry.Role.master) {
+                String lastNode = masters.putIfAbsent(entry.epoch, entry.node);
+                if (lastNode != null) {
+                    Assert.assertEquals(lastNode, entry.node);
+                }
+            }
+        }
+
+        Assert.assertTrue(masters.size() > 0);
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/RoleElectionStateMachineTest.java:322]: we can import util.Assert and use Assert.assertGt here

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java [hunk 26]
```diff
@@ -0,0 +1,310 @@
...
+        @Override
+        public Callback callback(StateMachineCallback callback) {
+            return callback::master;
+        }
+    }
+
+    private static class WorkerState implements RoleState {
+
+        private RoleStateData stateData;
+        private int count = 0;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java:186]: move assignment to the constructor

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java [hunk 27]
```diff
@@ -0,0 +1,310 @@
...
+
+import java.security.SecureRandom;
+import java.util.Optional;
+import java.util.concurrent.locks.LockSupport;
+
+import com.baidu.hugegraph.util.E;
+
+public class RoleElectionStateMachineImpl implements RoleElectionStateMachine {
+
+    private volatile boolean shutdown = false;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java:30]: move assignment to the constructor

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleStateData.java [hunk 28]
```diff
@@ -0,0 +1,91 @@
...
+    public void count(long count) {
+        this.count = count;
+    }
+
+    public String node() {
+        return this.node;
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleStateData.java:65]: o looks like 0, can rename to obj?

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/RoleElectionStateMachineTest.java [hunk 29]
```diff
@@ -0,0 +1,324 @@
...
+        }
+
+        public LogEntry(Integer epoch, String node, Role role) {
+            this.epoch = epoch;
+            this.node = node;
+            this.role = role;
+        }
+
+        @Override
+        public boolean equals(Object o) {
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/RoleElectionStateMachineTest.java:71]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java [hunk 30]
```diff
@@ -0,0 +1,310 @@
...
+            }
+        }
+
+        @Override
+        public Callback callback(StateMachineCallback callback) {
+            return callback::unknown;
+        }
+    }
+
+    private static class SafeState implements RoleState {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java:137]: rename class name

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java [hunk 31]
```diff
@@ -244,10 +246,10 @@ public RoleState transform(StateMachineContext context) {
             RoleStateData stateData = new RoleStateData(context.config().node(), epoch);
             //failover to master success
             context.epoch(stateData.epoch());
-            if (context.adapter().delayIfNodePresent(stateData, -1)) {
+            if (context.adapter().updateIfNodePresent(stateData)) {
                 return new MasterState(stateData);
             } else {
-                return new WorkerState(stateData);
+                return new UnKnownState(epoch).transform(context);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java:252]: UnKnown => Unknown

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java [hunk 32]
```diff
@@ -183,10 +184,11 @@ public Callback callback(StateMachineCallback callback) {
     private static class WorkerState implements RoleState {
 
         private RoleStateData stateData;
-        private int count = 0;
+        private int count;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleElectionStateMachineImpl.java:187]: also rename to clock?

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/RoleElectionStateMachineTest.java [hunk 33]
```diff
@@ -68,11 +68,12 @@ public LogEntry(Integer epoch, String node, Role role) {
         }
 
         @Override
-        public boolean equals(Object o) {
-            if (this == o) return true;
-            if (!(o instanceof LogEntry)) return false;
-            LogEntry logEntry = (LogEntry) o;
-            return Objects.equals(epoch, logEntry.epoch) && Objects.equals(node, logEntry.node) && role == logEntry.role;
+        public boolean equals(Object obj) {
+            if (this == obj) return true;
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/RoleElectionStateMachineTest.java:72]: add "{}" and wrap line

File: hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleTypeDataAdapter.java [hunk 34]
```diff
@@ -0,0 +1,29 @@
...
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.baidu.hugegraph.election;
+
+import java.util.Optional;
+
+public interface RoleTypeDataAdapter {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/election/RoleTypeDataAdapter.java:24]: keep the name style consistent: RoleTypeDataAdapter and RoleStateData

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/RoleElectionStateMachineTest.java [hunk 35]
```diff
@@ -0,0 +1,320 @@
...
+                    }
+
+                    Assert.assertEquals(value.epoch(), copy.epoch());
+                    if (Objects.equals(value.node(), copy.node()) &&
+                        value.clock() <= copy.clock()) {
+                        System.out.println("----2" + copy);
+                        metaDataLogs.add(copy);
+                        if (value.clock() == copy.clock()) {
+                            Exception e = new Exception("eq");
+                            e.printStackTrace();
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/RoleElectionStateMachineTest.java:240]: call assertFail?

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/RoleElectionStateMachineTest.java [hunk 36]
```diff
@@ -0,0 +1,320 @@
...
+                    return false;
+                }
+
+                RoleTypeData copy = this.copy(stateData);
+                RoleTypeData newData = data.compute(copy.epoch(), (key, value) -> {
+                    if (copy.epoch() > this.epoch) {
+                        this.epoch = copy.epoch();
+                        Assert.assertNull(value);
+                        metaDataLogs.add(copy);
+                        System.out.println("----1" + copy);
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/RoleElectionStateMachineTest.java:229]: improve print message

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/RoleElectionStateMachineTest.java [hunk 37]
```diff
@@ -0,0 +1,320 @@
...
+                        return copy;
+                    }
+
+                    Assert.assertEquals(value.epoch(), copy.epoch());
+                    if (Objects.equals(value.node(), copy.node()) &&
+                        value.clock() <= copy.clock()) {
+                        System.out.println("----2" + copy);
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/RoleElectionStateMachineTest.java:236]: ditto

**Review by javeme — APPROVED** [review]

**Review by javeme — APPROVED** [review]

**Review by imbajin — APPROVED** [review]

**Pull request merged**